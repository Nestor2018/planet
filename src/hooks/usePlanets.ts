import { useState, useEffect } from "react";

import { API_HOST } from "../utils/constants";

const API_URL = API_HOST;

/**
 * Custom hook to fetch and manage planet data.
 *
 * @returns {Object} An object containing the planets data, loading state, and error state.
 */
const usePlanets = () => {
  const [planets, setPlanets] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    const fetchPlanets = async () => {
      try {
        const response = await fetch(API_URL);
        if (!response.ok) throw new Error("Failed to fetch planets");
        const data = await response.json();
        setPlanets(data.bodies);
      } catch (err) {
        setError(err.message);
      } finally {
        setLoading(false);
      }
    };

    fetchPlanets();
  }, []);

  return { planets, loading, error };
};

export default usePlanets;
