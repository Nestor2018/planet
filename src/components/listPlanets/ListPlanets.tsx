import React, { useState, useEffect } from "react";
import { View, Text, StyleSheet } from "react-native";

import { PlanetType } from "../../interfaces/planet.interface";
import usePlanets from "../../hooks/usePlanets";
import colors from "../../utils/colors";
import Loader from "../loader/Loader";
import PlanetSearch from "../planetSearch/PlanetSearch";
import SortButtons from "./SortButtons";
import PlanetList from "./PlanetList";

const ListPlanets: React.FC = () => {
  const { planets, loading, error } = usePlanets() as {
    planets: PlanetType[];
    loading: boolean;
    error: string | null;
  };
  const [searchText, setSearchText] = useState("");
  const [sortOrder, setSortOrder] = useState<"asc" | "desc">("asc");
  const [filteredPlanets, setFilteredPlanets] = useState<PlanetType[]>([]);

  useEffect(() => {
    const updatedPlanets = planets
      .filter((planet) =>
        planet.englishName.toLowerCase().includes(searchText.toLowerCase()),
      )
      .sort((a, b) =>
        sortOrder === "asc"
          ? a.englishName.localeCompare(b.englishName)
          : b.englishName.localeCompare(a.englishName),
      );

    setFilteredPlanets(updatedPlanets);
  }, [planets, searchText, sortOrder]);

  if (error) return <Text style={styles.error}>{error}</Text>;

  return (
    <View style={styles.container}>
      {loading && <Loader testID="loader" />}
      <PlanetSearch onChange={setSearchText} />
      <SortButtons sortOrder={sortOrder} setSortOrder={setSortOrder} />
      <PlanetList planets={filteredPlanets} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.blackPearl,
    padding: 10,
  },
  error: {
    color: colors.carmine,
    textAlign: "center",
    marginTop: 20,
  },
});

export default ListPlanets;
