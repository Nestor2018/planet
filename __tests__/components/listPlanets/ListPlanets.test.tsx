import React from "react";
import { render, waitFor } from "@testing-library/react-native";
import ListPlanets from "../../../src/components/listPlanets/ListPlanets";
import usePlanets from "../../../src/hooks/usePlanets";

jest.mock("../../../src/hooks/usePlanets", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("ListPlanets Component", () => {
  const mockPlanets = [
    { englishName: "Earth", id: "earth" },
    { englishName: "Mars", id: "mars" },
    { englishName: "Jupiter", id: "jupiter" },
  ];

  beforeEach(() => {
    (usePlanets as jest.Mock).mockClear();
  });

  it("renders Loader when loading is true", () => {
    (usePlanets as jest.Mock).mockReturnValue({
      planets: [],
      loading: true,
      error: null,
    });

    const { getByTestId } = render(<ListPlanets />);
    expect(getByTestId("loader")).toBeTruthy();
  });

  it("renders error message when error exists", () => {
    (usePlanets as jest.Mock).mockReturnValue({
      planets: [],
      loading: false,
      error: "Failed to fetch planets",
    });

    const { getByText } = render(<ListPlanets />);
    expect(getByText("Failed to fetch planets")).toBeTruthy();
  });

  it("matches snapshot when loaded", () => {
    (usePlanets as jest.Mock).mockReturnValue({
      planets: mockPlanets,
      loading: false,
      error: null,
    });

    const tree = render(<ListPlanets />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
