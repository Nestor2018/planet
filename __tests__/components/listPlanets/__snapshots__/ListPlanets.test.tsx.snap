// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPlanets Component matches snapshot when loaded 1`] = `
<View
  style={
    {
      "backgroundColor": "#20252c",
      "flex": 1,
      "padding": 10,
    }
  }
>
  <View
    testID="planet-search"
  >
    <TextInput
      accessibilityLabel="Search planet"
      accessible={true}
      onChangeText={[Function]}
      placeholder="Search planet"
      placeholderTextColor="white"
      style={
        [
          {
            "backgroundColor": "#a0a0a0",
            "color": "#fff",
            "height": 46,
            "paddingLeft": 16,
          },
          {
            "borderRadius": 8,
            "margin": 8,
          },
        ]
      }
      testID="input"
      value=""
    />
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "center",
        "marginVertical": 10,
      }
    }
    testID="sort-buttons"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#3c6fc8",
          "borderRadius": 8,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginHorizontal": 5,
          "minWidth": 120,
          "opacity": 1,
          "padding": 10,
        }
      }
      testID="button-asc"
    >
      <view
        color="#20252c"
        name="arrow-up"
        size={18}
        testID="icon-arrow-up"
      />
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontSize": 14,
              "marginLeft": 5,
            },
            {
              "color": "#20252c",
            },
          ]
        }
      >
        Ascendente
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#272c35",
          "borderRadius": 8,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginHorizontal": 5,
          "minWidth": 120,
          "opacity": 1,
          "padding": 10,
        }
      }
      testID="button-desc"
    >
      <view
        color="#fff"
        name="arrow-down"
        size={18}
        testID="icon-arrow-down"
      />
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontSize": 14,
              "marginLeft": 5,
            },
            false,
          ]
        }
      >
        Descendente
      </Text>
    </View>
  </View>
  <RCTScrollView
    ListEmptyComponent={
      <Text
        style={
          {
            "color": "#fff",
            "fontSize": 16,
            "marginTop": 16,
          }
        }
      >
        No planets available.
      </Text>
    }
    data={
      [
        {
          "englishName": "Earth",
          "id": "earth",
        },
        {
          "englishName": "Jupiter",
          "id": "jupiter",
        },
        {
          "englishName": "Mars",
          "id": "mars",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    testID="planet-list"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#272c35",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginVertical": 5,
                "padding": 10,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontSize": 16,
                }
              }
            >
              Earth
            </Text>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <view
                color="#fff"
                name="heart-outline"
                size={24}
                testID="icon-heart-outline"
              />
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#272c35",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginVertical": 5,
                "padding": 10,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontSize": 16,
                }
              }
            >
              Jupiter
            </Text>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <view
                color="#fff"
                name="heart-outline"
                size={24}
                testID="icon-heart-outline"
              />
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#272c35",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginVertical": 5,
                "padding": 10,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontSize": 16,
                }
              }
            >
              Mars
            </Text>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <view
                color="#fff"
                name="heart-outline"
                size={24}
                testID="icon-heart-outline"
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
